src/backend/distributed/planner/intermediate_result_pruning.c

```diff
@@ -150,9 +149,11 @@ RecordSubplanExecutionsOnNodes(HTAB *intermediateResultsHash,
  * the workers will be in the resulting list.
  */
 static List *
-AppendAllAccessedWorkerNodes(List *workerNodeList, DistributedPlan *distributedPlan, int
+AppendAllAccessedWorkerNodes(IntermediateResultsHashEntry *entry,
+							 DistributedPlan *distributedPlan, int
 							 workerNodeCount)
 {
+	List *workerNodeList = entry->nodeIdList;
```

Comment:
We capture `entry->nodeIdList` in `workerNodeList` and during the runtime of this function we mutate `workerNodeList`.

The comment for `lappend` and its family of functions states:
```C
/*
 * Append a pointer to the list. A pointer to the modified list is
 * returned. Note that this function may or may not destructively
 * modify the list; callers should always use this function's return
 * value, rather than continuing to use the pointer passed as the
 * first argument.
 */
List *
lappend(List *list, void *datum)
```
source: https://github.com/postgres/postgres/blob/41c184bc642b25f67fb1d8ee290f28805fa5a0b4/src/backend/nodes/list.c#L314-L322

We leave the original pointer on the `entry->nodeIdList` after we return control back to the caller.

The first thing the caller does now is put the return value (`workerNodeList` back on `entry->nodeIdList`) so there is no direct bug, however the code flow in its current form makes this hard to verify.

Instead I would suggest to make this functions return type a void and make sure to append the nodeId's directly to `entry->nodeIdList`.

Even though it is not introduced in this patch I think it is a good moment to move the mutation of `entry->nodeIdList` into this function to prevent hard to catch bugs from wrong use of this function.